File: 0242-valid-anagram/0242-valid-anagram.py
class Solution:
    def isAnagram(self, s: str, t: str) -> bool:
        if len(s)!=len(t):
            return False
        
        mp = dict()
        
        for c in s:
            mp[c] = 1+mp.get(c,0)
            
        for c in t:
            if c not in mp or (c in mp and mp[c] == 0):
                return False
            
            mp[c]-=1
            
            if mp[c] == 0:
                del mp[c]
        
        # print(mp)
        
        return True if not mp else False

File: 0242-valid-anagram/0242-valid-anagram.cpp
class Solution {
public:
    bool isAnagram(string s, string t) {
        unordered_map<char, int> mp;
        
        for(auto &c:s){
            auto rec = mp.find(c);
            if(rec == mp.end()){
                mp.insert(make_pair(c,1));
            }
            else{
                mp[c]+=1;
            }
        }
        
        for(auto &c:t){
            auto rec = mp.find(c);
            if(rec == mp.end()) return false;
            
            if(rec != mp.end() and mp[c] > 0){
                mp[c]-=1;
                
                if(mp[c]==0){
                    mp.erase(c);
                }
            }
        }
        
        if(mp.size() > 0){
            return false;
        }
        
        return true; 
    }
};